src/Resource/Collection.js

```diff
@@ -1,15 +1,15 @@
 import ParameterBag from './ParameterBag';
 
-/**
- * @param {String} name
- * @param {HyralRepository} repository
- */
-function Collection(name, repository) {
-  if (!new.target) {
-    throw Error('Collection() must be called with new');
-  }
-
-  Object.defineProperties(this, {
+function defineCollectionProperties(collection, {
```

Comment:
ik twijfel over de losse variables voor metadata. Zouden we niet gewoon metadata in zijn geheel moeten meegeven in deze functie?

Deze constructie zet mij verder aan het denken: je gebruikt nu een semi-factory functie. Waarom passen we de API niet gewoon aan naar een complete factory implementatie? const collection = createCollection(). Dan geen Object.create() in clone maar gewoon createCollection.

Is Object.create(obj.prototype) niet gewoon hetzelfde (effectief) als new Obj()?

Reply:
Ik heb er nog een beetje over nagedacht en ik ben het eens dat die semi-factory functie niet logisch is. 

Willen we bepaalde data echt private houden? In de huidige implementatie kun je namelijk de items gewoon overschrijven door `collection.data.items` aan te passen. Alleen `collection.data = ...` kan niet. De data referentie kan namelijk niet meer veranderen, maar je kunt in het data object gerust dingen veranderen. 

Als we private data willen hebben neig ik toch naar een constructor function, waarbij je optioneel constructie data mee kunt geven die je uit de store haalt. Na initialisatie zijn dan de data die we private willen houden nog steeds private.

```
function Collection(name, initialPrivate) {
  const privateData = initialPrivate || {
    items: [],
    invisible: name,
  };

  this.name = name;

  Object.defineProperties(this, {
    items: {
      get: () => privateData.items,
      enumerable: true,
    },
  });
}
```
De private data zijn dan eventueel met een getter property weer op te halen, maar ze zijn niet te wijzigen als wij dat niet willen.

Reply:
Hmm, interessant. We neigen toch al naar een constructor function en op deze manier houden we wel iets aan de nodige controle voor een aantal van de gewenste features. Ik stem voor privateData.

Reply:
Ga ik daar mee aan de slag :-)

Reply:
Ik heb nu een vrij radicale versie gemaakt zonder prototype.
Dit betekent dat eigenlijk alle data in de closure van de constructie functie zit, waardoor alles in principe private is. Ik definieer vervolgens alle getters en methods in het object zelf.
Het grote onderscheid tussen privateData en serializebleData is dat privateData volatile is: het is niet nodig dat deze data geserialiseerd wordt.